cmd/cluster/dynatrace/hcpMgCmd.go

```diff
@@ -0,0 +1,302 @@
+	hcpMgCmd.Flags().StringVar(&sortOrder, "sort", "desc", "Sort the results by timestamp in either ascending or descending order. Accepted values are 'asc' and 'desc'")
+
+	return hcpMgCmd
+}
+
+func mustGather(clusterID string) (error error) {
+	accessToken, err := getAccessToken()
+	if err != nil {
+		return fmt.Errorf("failed to acquire access token %v", err)
+	}
+
+	clusterInternalID, managementClusterName, DTURL, err := fetchClusterDetails(clusterID)
+	if err != nil {
+		return err
+	}
+	managementClusterInternalID, _, _, err := fetchClusterDetails(managementClusterName)
+	if err != nil {
+		return err
+	}
+
+	connection, err := getConnection()
+	if err != nil {
+		return err
+	}
+	clientset, err := getClientsetFromClusterID(connection, managementClusterInternalID)
+	if err != nil {
+		return err
+	}
+
+	hcpNS, err := GetHCPNamespaceFromInternalID(clientset, clusterInternalID)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println(fmt.Sprintf("Using HCP Namespace %v", hcpNS))
+
+	gatherDir, err := setupGatherDir(hcpNS)
+	if err != nil {
+		return err
+	}
+
+	pods, err := getPodsForNamespace(clientset, hcpNS)
+	if err != nil {
+		return err
+	}
+
+	err = dumpPodLogs(pods, gatherDir, hcpNS, managementClusterName, DTURL, accessToken, since, tail, sortOrder)
+	if err != nil {
+		return err
+	}
+
+	deployments, err := getDeploymentsForNamespace(clientset, hcpNS)
+	if err != nil {
+		return err
+	}
+
+	err = dumpEvents(deployments, gatherDir, hcpNS, managementClusterName, DTURL, accessToken, since, tail, sortOrder)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func dumpEvents(deploys *appsv1.DeploymentList, gatherDir string, hcpNS string, managementClusterName string, DTURL string, accessToken string, since int, tail int, sortOrder string) error {
+	totalDeployments := len(deploys.Items)
+	for k, d := range deploys.Items {
+		fmt.Println(fmt.Sprintf("[%d/%d] Deployment events for %s", k+1, totalDeployments, d.Name))
+
+		eventQuery, err := getEventQuery(d.Name, hcpNS, since, tail, sortOrder, managementClusterName)
+		if err != nil {
+			return err
+		}
+		eventQuery.Build()
+
+		eventsRequestToken, err := getRequestToken(eventQuery.finalQuery, DTURL, accessToken)
+		if err != nil {
+			fmt.Println(fmt.Errorf("failed to acquire request token %v", err))
+			continue
+		}
+
+		eventsDirPath, err := addEventsDir(gatherDir, d.Name)
+		if err != nil {
+			return err
+		}
+
+		deploymentYamlPath := filepath.Join(eventsDirPath, "deployment.yaml")
+		deploymentYaml, err := yaml.Marshal(d)
+		f, err := os.OpenFile(deploymentYamlPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0655)
+		if err != nil {
+			return err
+		}
+		f.Write(deploymentYaml)
+		f.Close()
+
+		eventsFilePath := filepath.Join(eventsDirPath, "events.log")
+		f, err = os.OpenFile(eventsFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0655)
+		if err != nil {
+			return err
+		}
+
+		err = getEvents(DTURL, accessToken, eventsRequestToken, f)
+		defer f.Close()
```

Comment:
Not something to hold the PR up for, but it's generally considered risky to put a `defer` in a for-loop because the resources won't be released until the entire loop is completed.

Some workarounds I've seen for this include wrapping the code in an anonymous function, or creating a new named function to call.